File: test/artifacts.test.mjs
/**
 * Answer artifacts, TS twin (src/server/vault.ts::writeArtifact): name repair
 * and collision safety MUST mirror native vault.rs::write_artifact
 * (openspec: add-answer-artifacts).
 *
 * Run: `node --test test/artifacts.test.mjs`
 */
import { test } from "node:test";
import assert from "node:assert/strict";
import { register } from "node:module";
import { existsSync, mkdtempSync, mkdirSync, readFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

register("./_ts-extensionless-hook.mjs", import.meta.url);

// Point the vault at a temp dir BEFORE importing the server modules.
const home = mkdtempSync(path.join(tmpdir(), "lh-artifact-"));
const vault = path.join(home, "vault");
mkdirSync(vault, { recursive: true });
process.env.VAULT_DIR = vault;

const { writeArtifact } = await import("../src/server/vault.ts");

test("hostile hints are repaired and stay inside the vault", () => {
  const { id, name } = writeArtifact("Lighthouse Notes", "../../evil name", "md", Buffer.from("x"));
  assert.equal(id, `Lighthouse Notes/${name}`);
  const abs = path.join(vault, id);
  assert.ok(existsSync(abs), `written inside the vault: ${abs}`);
  assert.ok(!existsSync(path.join(home, "evil name.md")), "no traversal outside the vault");

  // An empty / dotfile hint falls back instead of failing.
  const fallback = writeArtifact("Lighthouse Notes", "...", "md", Buffer.from("y"));
  assert.equal(fallback.name, "result.md");
});

test("collisions suffix; existing artifacts are never overwritten", () => {
  const a = writeArtifact("Lighthouse Results", "totals", "csv", Buffer.from("a"));
  const b = writeArtifact("Lighthouse Results", "totals", "csv", Buffer.from("b"));
  assert.equal(a.name, "totals.csv");
  assert.equal(b.name, "totals (1).csv");
  assert.equal(readFileSync(path.join(vault, "Lighthouse Results/totals.csv"), "utf8"), "a");
  assert.equal(readFileSync(path.join(vault, "Lighthouse Results/totals (1).csv"), "utf8"), "b");
});

File: native/crates/lighthouse-core/tests/artifacts_test.rs
//! Answer artifacts over a real temp vault: write_artifact name repair +
//! collision safety, and the Save-as-CSV path end to end
//! (openspec: add-answer-artifacts).

mod common;

use lighthouse_core::analytics::run_direct_save;
use lighthouse_core::vault::{self, write_artifact};

fn write(path: &std::path::Path, text: &str) {
    std::fs::create_dir_all(path.parent().unwrap()).unwrap();
    std::fs::write(path, text).unwrap();
}

#[test]
fn write_artifact_repairs_names_and_never_overwrites() {
    let dir = tempfile::tempdir().unwrap();
    let _guard = common::lock_env(dir.path());

    // A hostile hint is repaired, not rejected — and stays inside the vault.
    let (id, name) = write_artifact("Lighthouse Notes", "../../evil name", "md", b"x").unwrap();
    assert_eq!(id, format!("Lighthouse Notes/{name}"));
    assert!(name.starts_with("..-..-evil name") || name.contains("evil name"), "{name}");
    let abs = dir.path().join(&id);
    assert!(abs.exists(), "written inside the vault: {abs:?}");

    // An empty / dotfile hint falls back instead of failing.
    let (_, fallback) = write_artifact("Lighthouse Notes", "...", "md", b"y").unwrap();
    assert_eq!(fallback, "result.md");

    // Same hint again ⇒ collision suffix, both files intact.
    let (_, first) = write_artifact("Lighthouse Results", "totals", "csv", b"a").unwrap();
    let (_, second) = write_artifact("Lighthouse Results", "totals", "csv", b"b").unwrap();
    assert_eq!(first, "totals.csv");
    assert_eq!(second, "totals (1).csv");
    assert_eq!(
        std::fs::read(dir.path().join("Lighthouse Results/totals.csv")).unwrap(),
        b"a"
    );
}

#[tokio::test]
async fn save_as_csv_writes_a_queryable_vault_file() {
    let vault_dir = tempfile::tempdir().unwrap();
    let _guard = common::lock_env(vault_dir.path());
    write(
        &vault_dir.path().join("sales.csv"),
        "region,amount\nNE,100\nNW,50\nNE,25\n",
    );
    vault::invalidate_walk_cache();
    // Direct execution honors AI visibility — only included files register.
    vault::set_included("sales.csv", true);

    let ids = vec!["sales.csv".to_string()];
    let (preview, saved) = run_direct_save(
        "SELECT region, SUM(amount) AS total FROM sales GROUP BY region ORDER BY total DESC",
        &ids,
        "Totals by region",
    )
    .await
    .expect("save runs");

    // The preview carries the normal narration + provenance.
    assert!(preview.markdown.contains("NE"), "{}", preview.markdown);
    assert!(preview.footer.contains("*Query used:*"), "{}", preview.footer);

    // The artifact is a real, full-fidelity CSV in Lighthouse Results/.
    assert_eq!(saved.name, "Totals by region.csv");
    assert_eq!(saved.rows, 2);
    let csv = std::fs::read_to_string(
        vault_dir.path().join("Lighthouse Results").join(&saved.name),
    )
    .unwrap();
    assert_eq!(csv, "region,total\r\nNE,125\r\nNW,50\r\n");

    // And the walk sees it immediately — it's ordinary vault input now.
    let nodes = vault::list_nodes();
    assert!(
        nodes.iter().any(|n| n.id == saved.id),
        "saved artifact appears in the tree"
    );

    // A write statement is still rejected on the save path.
    let err = run_direct_save("DROP TABLE sales", &ids, "nope").await.unwrap_err();
    assert!(err.contains("SELECT"), "{err}");
    assert!(!vault_dir.path().join("Lighthouse Results/nope.csv").exists());
}
